src/flash.rs

```diff
@@ -217,29 +253,60 @@ impl<'a, const SECTOR_SZ_KB: u32> FlashWriter<'a, SECTOR_SZ_KB> {
         )
     }
 
-    /// Write data to `FLASH_START + offset`
-    pub fn write(&mut self, offset: u32, data: &[u8]) -> Result<()> {
-        self.valid_length(offset, data.len())?;
+    /// Write data to `FLASH_START + offset`.
+    ///
+    /// If `force_data_padding` is true, the incoming data will be padded with 0xFF
+    /// to satisfy the requirement that 2 words (8 bytes) must be written at a time.
+    /// If `force_data_padding` is false and `data.len()` is not divisible by 8,
+    /// the error `ArrayMustBeDivisibleBy8` will be returned.
+    pub fn write(&mut self, offset: u32, data: &[u8], force_data_padding: bool) -> Result<()> {
+        self.valid_length(offset, data.len(), force_data_padding)?;
 
         // Unlock Flash
         self.unlock()?;
 
-        for idx in (0..data.len()).step_by(2) {
+        // According to RM0440 Rev 7, "It is only possible to program double word (2 x 32-bit data)"
+        for idx in (0..data.len()).step_by(8) {
+            // Check the starting address
             self.valid_address(offset + idx as u32)?;
+            // Check the ending address to make sure there is no overflow
+            self.valid_address(offset + 8 + idx as u32)?;
 
-            let write_address = (FLASH_START + offset + idx as u32) as *mut u16;
+            let write_address1 = (FLASH_START + offset + idx as u32) as *mut u32;
+            let write_address2 = (FLASH_START + offset + 4 + idx as u32) as *mut u32;
+
+            let word1: u32;
+            let word2: u32;
+
+            // Check if there is enough data to make 2 words, if there isn't, pad the data with 0xFF
+            if idx + 8 > data.len() {
+                let mut tmp_buffer = [255 as u8; 8];
```

Comment:
should be `255u8`?

Reply:
also, why this buffer length use?

Reply:
That buffer is used to pad non-divisible-by-8 data if the user wants an "easy" way to write data to flash. The STM32G4 requires 8 bytes to be written at a time. The user can use the `force_data_padding` flag so that data that isn't divisible by 8 is buffered into `tmp_buffer` and appropriately padded on the fly. 

I can change it to 255u8 if that is preferred.

Reply:
Great, also don't forget remove TODO for flash register structure.

Reply:
Ok, I think it is complete, thank you for reviewing!